HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,20 @@
+from pyspark import SparkContext
+from pyspark import AccumulatorParam
+
+# set accumulator to keep track of visited nodes
+class SetAccumulatorParam(AccumulatorParam):
+    def zero(self, initialValue):
+        return set()
+    def addInPlace(self, v1, v2):
+        return v1 | v2
+
+def rdd_bfs(root, graph, sc):
+    visited = sc.accumulator(set(), SetAccumulatorParam());
+    neigh_set = set([root])
+    while neigh_set:
+        curr_visited = visited.value
+        # not doing any calls to collect, only using the accumulator
+        graph.filter(lambda (k, v): k in neigh_set).foreach(lambda (k, v): visited.add(set(v)))

```

Comment:
I am not sure if the filter is in-place, but an optimization will be to filter out the neigh_set from the graph, so that you have decreasing number of nodes to work with